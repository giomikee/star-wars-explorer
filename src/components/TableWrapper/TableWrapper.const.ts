export const TABLE_WRAPPER_EVENTS = {
  pageChanged: 'page-changed',
} as const
