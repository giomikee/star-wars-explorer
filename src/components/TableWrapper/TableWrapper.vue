<template>
  <v-container>
    <v-skeleton-loader :loading="isLoading" type="table">
      <slot />
      <v-pagination v-model="tableCurrentPage" class="w-100" :length="pages" />
    </v-skeleton-loader>
  </v-container>
</template>

<script lang="ts" setup>
  import { ref, toRef, watch } from 'vue'
  import { TABLE_WRAPPER_EVENTS } from './TableWrapper.const'

  interface Props {
    currentPage: number
    isLoading: boolean
    pages: number
  }

  const props = defineProps<Props>()

  const emit = defineEmits(Object.values(TABLE_WRAPPER_EVENTS))

  const tableCurrentPage = ref<number>(props.currentPage)
  const currentPageProp = toRef(props, 'currentPage')

  watch(currentPageProp, () => {
    tableCurrentPage.value = currentPageProp.value
  })

  watch(tableCurrentPage, () => {
    emit(TABLE_WRAPPER_EVENTS.pageChanged, tableCurrentPage.value)
  })

</script>

<script lang="ts">
  export default { name: 'TableWrapper' }
</script>
