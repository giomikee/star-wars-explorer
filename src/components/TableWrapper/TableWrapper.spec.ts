import { shallowMount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import { nextTick } from 'vue'
import { TABLE_WRAPPER_EVENTS } from './TableWrapper.const'
import TableWrapper from './TableWrapper.vue'

interface Component {
  tableCurrentPage: number
}

const createWrapper = (props: { currentPage: number, isLoading: boolean, pages: number }) => shallowMount(
  TableWrapper,
  {
    props,
    global: {
      stubs: {
        VSkeletonLoader: false,
      },
    },
  },
)

describe('Given TableWrapper component', () => {
  describe.each([true, false])('When component is rendered', isLoading => {
    describe(`And isLoading is ${isLoading}`, () => {
      it('Then it should match snapshot', () => {
        const { element } = createWrapper({
          currentPage: 1,
          isLoading,
          pages: 2,
        })

        expect(element).toMatchSnapshot()
      })
    })
  })

  describe('Given component watchers', () => {
    describe('When currentPage prop is set to 2', () => {
      it('Then tableCurrentPage should be set to 2', async () => {
        const wrapper = createWrapper({
          currentPage: 1,
          isLoading: false,
          pages: 5,
        })

        await wrapper.setProps({ currentPage: 2 })

        expect((wrapper.vm as typeof wrapper.vm & Component).tableCurrentPage).toBe(2)
      })
    })

    describe('When tableCurrentPage is updated to 2', () => {
      it(`Then "${TABLE_WRAPPER_EVENTS.pageChanged}" event should be emitted with 2 as payload `, async () => {
        const wrapper = createWrapper({
          currentPage: 1,
          isLoading: false,
          pages: 5,
        });

        (wrapper.vm as typeof wrapper.vm & Component).tableCurrentPage = 2

        await nextTick()

        expect(wrapper.emitted()[TABLE_WRAPPER_EVENTS.pageChanged]).toEqual([[2]])
      })
    })
  })
})
