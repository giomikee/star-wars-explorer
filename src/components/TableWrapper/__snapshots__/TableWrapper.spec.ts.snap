// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given TableWrapper component > When component is rendered > And isLoading is false > Then it should match snapshot 1`] = `
<div
  class="v-container v-locale--is-ltr"
>
  <div
    class="v-skeleton-loader v-theme--dark"
  >
    
    
    
    <v-pagination-stub
      aria-label="$vuetify.pagination.ariaLabel.root"
      border="false"
      class="w-100"
      currentpagearialabel="$vuetify.pagination.ariaLabel.currentPage"
      density="default"
      disabled="false"
      ellipsis="..."
      firstarialabel="$vuetify.pagination.ariaLabel.first"
      firsticon="$first"
      lastarialabel="$vuetify.pagination.ariaLabel.last"
      lasticon="$last"
      length="2"
      modelvalue="1"
      nextarialabel="$vuetify.pagination.ariaLabel.next"
      nexticon="$next"
      pagearialabel="$vuetify.pagination.ariaLabel.page"
      previcon="$prev"
      previousarialabel="$vuetify.pagination.ariaLabel.previous"
      showfirstlastpage="false"
      size="default"
      start="1"
      tag="nav"
      tile="false"
      variant="text"
    />
    
  </div>
</div>
`;

exports[`Given TableWrapper component > When component is rendered > And isLoading is true > Then it should match snapshot 1`] = `
<div
  class="v-container v-locale--is-ltr"
>
  <div
    aria-label="Loading..."
    aria-live="polite"
    class="v-skeleton-loader v-theme--dark"
    role="alert"
  >
    
    <div
      class="v-skeleton-loader__bone v-skeleton-loader__table"
    >
      
      
      <div
        class="v-skeleton-loader__bone v-skeleton-loader__table-heading"
      >
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__chip"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__text"
        >
          
          
        </div>
        
        
      </div>
      
      
      <div
        class="v-skeleton-loader__bone v-skeleton-loader__table-thead"
      >
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__heading"
        >
          
          
        </div>
        
        
      </div>
      
      
      <div
        class="v-skeleton-loader__bone v-skeleton-loader__table-tbody"
      >
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__table-row-divider"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__table-row"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__divider"
          >
            
            
          </div>
          
          
        </div>
        
        
      </div>
      
      
      <div
        class="v-skeleton-loader__bone v-skeleton-loader__table-tfoot"
      >
        
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__text"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__text"
        >
          
          
        </div>
        
        
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__avatar"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__avatar"
        >
          
          
        </div>
        
        
        
      </div>
      
      
    </div>
    
  </div>
</div>
`;
